ERCS/erc-7527.md

```diff
@@ -0,0 +1,510 @@
+     * @param data The data to encode into ifself and the newly created non-fungible token.
+     * @return The identifier of the newly created non-fungible token.
+     */
+    function wrap(address to, bytes calldata data) external payable returns (uint256);
+
+    /**
+     * @dev Unwrap a non-fungible token into currency premium.
+     *
+     * Todo: event
+     *
+     * @param to The address of the recipient of the currency.
+     * @param tokenId The identifier of the non-fungible token to unwrap.
+     * @param data The data to encode into ifself and the non-fungible token with identifier `tokenId`.
+     */
+    function unwrap(address to, uint256 tokenId, bytes calldata data) external payable;
+
+    /**
+     * @dev Returns the strategy of the agency.
+     * @return app The address of the app.
+     * @return asset The asset of the agency.
+     * @return attributeData The attributeData of the agency.
+     */
+    function getStrategy() external view returns (address app, Asset memory asset, bytes memory attributeData);
+
+    /**
+     * @dev Returns the price and fee of unwrapping.
+     * @param data The data to encode to calculate the price and fee of unwrapping.
+     * @return price The price of wrapping.
+     * @return fee The fee of wrapping.
+     */
+    function getUnwrapOracle(bytes memory data) external view returns (uint256 price, uint256 fee);
+
+    /**
+     * @dev Returns the price and fee of wrapping.
+     * @param data The data to encode to calculate the price and fee of wrapping.
+     * @return price The price of wrapping.
+     * @return fee The fee of wrapping.
+     */
+    function getWrapOracle(bytes memory data) external view returns (uint256 price, uint256 fee);
+}
+
+```
+
+### App Interface
+
+`ERC7527App` SHALL inherit `name` from interface `ERC721Metadata`. 
+
+``` 
+
+pragma solidity ^0.8.20;
+
+interface IERC7527App {
+    /**
+     * @dev Returns the maximum supply of the non-fungible token.
+     */
+    function getMaxSupply() external view returns (uint256);
+
+    /**
+     * @dev Returns the name of the non-fungible token with identifier `id`.
+     * @param id The identifier of the non-fungible token.
+     */
+    function getName(uint256 id) external view returns (string memory);
+
+    /**
+     * @dev Returns the agency of the non-fungible token.
+     */
+    function getAgency() external view returns (address payable);
+
+    /**
+     * @dev Sets the agency of the non-fungible token.
+     * @param agency The agency of the non-fungible token.
+     */
+    function setAgency(address payable agency) external;
+
+    /**
+     * @dev Mints a non-fungible token to `to`.
+     * @param to The address of the recipient of the newly created non-fungible token.
+     * @param data The data to encode into the newly created non-fungible token.
+     */
+    function mint(address to, bytes calldata data) external returns (uint256);
```

Comment:
1. Some prior ERCs addressed the minting and burning, such as https://eips.ethereum.org/EIPS/eip-5679, consider use them or add in rationale section the reasons for proposing a duplicate mint/burn purpose.
2. is it intentional that during minting the tokenId being minted is unknown / unspecified and can ONLY be unknown?

Reply:
1. updated in Rationale: 
    > [ERC-5679](./eip-5679.md) proposed `IERC5679Ext721` interface for introducing a consistent way to extend [ERC-721](./eip-721.md) token standards for minting and burning. To ensure the backward compatibility, considering some contracts which do not implement `ERC721TokenReceiver`, `IERC7527App` employ `mint` function instead of `safeMint`. To ensure the safety and the uniqueness of mutual bound, the `_from` parameter of the `burn` function in `IERC5679Ext721` must be the contract address of the bounded angency. Thus, `burn` function in `IERC7527App` does not contain the `_from` parameter. 

2. Token IDs are usually determined automatically by contracts. Here the `mint` function in `IERC7527App`, token ID can be passed by the parameter `data`.